### DISCUSSION TITLE
Embiggen or Ensmallen Selections

### DISCUSSION THREAD

**Discussion by summeremacs** [meow-edit/meow#314] [Q&A] [answered]
I have another question (sorry again...):

When I select a paragraph with meow (like . p ) or any other selection, what if I want to make the selection smaller or bigger character by character? I've noticed that pressing any keys afterwards for selection will deselect what you've selected. Is there a way to grow or shrink a selection, char by char, or word by word, after you've selected something? Or, for instance, when I'm using 'x' and I overshoot something...is there a way to go back with the selection instead of starting over?

I didn't see anything about it in the tutorial or the docs - I've probably missed something.

**Comment by eshrh** [comment, marked as answer]
There is currently no general meow way to grow a selection. I'd love to hear any elegant idea you might have.

However, for common situations, there are convenient solutions.
1. To grow a selection by characters -  `meow-left-expand` and `meow-right-expand` (bound to H and L iirc)
2. To grow vertically - `meow-next-expand` and `meow-prev-expand` (J and K)
3. If you have a selection by words (or lines), then you can always press `1` to expand by one more "thing" (as long you haven't pressed anything else)

It's useful to realize that meow is really very thin over plain emacs. A meow selection is really just an emacs "region" which is delimited by the "mark" and "point" (the cursor). Therefore, all plain emacs commands that manipulate the point/mark also work! So, to expand a word to the right even after pressing some keys, or from a sentence-based selection, you can just press `C-<right>` and `C-<left>`, which call built-in commands `left-word` and `right-word`. Similarly `C-a` `C-e` (line) `M-a` `M-e` (paragraph) and probably many many others also work. I'd refer to some standard emacs movement cheat sheet or the manual to find more.

> Or, for instance, when I'm using 'x' and I overshoot something...is there a way to go back with the selection instead of starting over?

This is a common issue. Call `meow-pop-selection`, i think it's bound to `z`. Do note that if you go outside meow with standard point movement commands then we can't keep track of selection history though.

**Reply by summeremacs** [reply]
Thank you!

Well, 'z' seems to do the trick for the most part. I did use it as "pop" in the tutorial but it didn't phase me what it was for. I thought it was just to pop back to the starting place so I sorta forgot about it (from the tutorial, I think that's all you do after selecting the buffer to the top and then to the bottom - pop back to the starting place). It doesn't explain in the tutorial that you can use it to deselect in steps. Somebody needs to just put in that explanation, I think (unless I missed it? It could be. I'm literally blonde).

EDIT: I misread your post! (I told you I'm blonde) I totally didn't make the distinction between J and K as opposed to j and k. This TOTALLY works, yes! ty! I don't remember it from the tutorial but yes it totally works ty. I hope somebody can add to the tutorial sometime to make these things a little more clear (I loved the tutorial but it seems like I missed a lot or didn't understand it enough which is probably on me). Please forgive me for my beginner level lack of comprehension. 🙃

**Reply by DogLooksGood** [reply]
BTW, you don't have to hold shift for `H` or `J`. Once you hit `H` or `J` once, the selection type will be updated to `char`, then `h` and `j` will expand the region.

**Reply by summeremacs** [reply]
tysm 💕 That's awesome I didn't even know that. I'm totally loving meow btw. ty guys for making it. 🙃

**Reply by quarkw** [reply]
A little detail needed for the first two suggestions:
1. To grow a selection by characters - meow-left-expand and meow-right-expand (bound to H and L iirc)
2. To grow vertically - meow-next-expand and meow-prev-expand (J and K)

Between the two expands you need to reverse the selection (`;` in the qwerty keymap) 

You'll also need to be aware of which direction the selection is. If the point is at the end of the selection, you'll actually need to reverse the order of the expand commands.

There could be a helper command that determines the selection direction, and calling the expand region commands for you in the correct order, and also calling `meow-reverse` twice, so your direction stays the same before/after the command
